pkg/util/runtime/runtime.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"runtime"
+)
+
+// For testing, bypass HandleCrash.
+var ReallyCrash bool
+
+// PanicHandlers is a list of functions which will be invoked when a panic happens.
+var PanicHandlers = []func(interface{}){logPanic}
+
+//TODO search the public functions
+// HandleCrash simply catches a crash and logs an error. Meant to be called via defer.
+// Additional context-specific handlers can be provided, and will be called in case of panic
+func HandleCrash(additionalHandlers ...func(interface{})) {
+	if ReallyCrash {
+		return
+	}
+	if r := recover(); r != nil {
+		for _, fn := range PanicHandlers {
+			fn(r)
+		}
+		for _, fn := range additionalHandlers {
+			fn(r)
+		}
+	}
+}
+
+// logPanic logs the caller tree when a panic occurs.
+func logPanic(r interface{}) {
+	callers := ""
+	for i := 0; true; i++ {
+		_, file, line, ok := runtime.Caller(i)
+		if !ok {
+			break
+		}
+		callers = callers + fmt.Sprintf("%v:%v\n", file, line)
+	}
+	glog.Errorf("Recovered from panic: %#v (%v)\n%v", r, r, callers)
+}
+
+// ErrorHandlers is a list of functions which will be invoked when an unreturnable
+// error occurs.
+var ErrorHandlers = []func(error){logError}
+
+// HandlerError is a method to invoke when a non-user facing piece of code cannot
+// return an error and needs to indicate it has been ignored. Invoking this method
+// is preferable to logging the error - the default behavior is to log but the
+// errors may be sent to a remote server for analysis.
+func HandleError(err error) {
```

Comment:
Sure, but we should maybe spread the word on when and how to use this.  I
didn't know it existed, and I still don't think I know how to apply it.

On Wed, Jan 20, 2016 at 2:48 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/util/runtime/runtime.go
> <https://github.com/kubernetes/kubernetes/pull/19778#discussion_r50333611>
> :
>
> > +			break
> > +		}
> > +		callers = callers + fmt.Sprintf("%v:%v\n", file, line)
> > +	}
> > +	glog.Errorf("Recovered from panic: %#v (%v)\n%v", r, r, callers)
> > +}
> > +
> > +// ErrorHandlers is a list of functions which will be invoked when an unreturnable
> > +// error occurs.
> > +var ErrorHandlers = []func(error){logError}
> > +
> > +// HandlerError is a method to invoke when a non-user facing piece of code cannot
> > +// return an error and needs to indicate it has been ignored. Invoking this method
> > +// is preferable to logging the error - the default behavior is to log but the
> > +// errors may be sent to a remote server for analysis.
> > +func HandleError(err error) {
>
> @thockin <https://github.com/thockin> either way, it's a big code smell
> (something is probably wrong if one's code "can't return an error"). But
> this is strictly better than a log message everywhere, because it's easy to
> find all the places that use it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19778/files#r50333611>.
>